### PR TITLE
Enabling eltwise TPPs and updating other libxsmm APIs to v2 for seismic part

### PR DESCRIPTION
**Pull request by kvoronin-intel**
...to be added

### CODE REVIEW

File: src/data/BinaryXsmm.hpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
+/**
+ * @file This file is part of EDGE.
+ *
+ * @author Alexander Breuer (anbreuer AT ucsd.edu)
```
Comment (alheinecke) [inline src/data/BinaryXsmm.hpp]: nit: please make you author and remove the other two, also please adjust the copyright

File: src/data/MmXsmmFused.hpp [hunk 2]
```diff
@@ -102,24 +95,23 @@ class edge::data::MmXsmmFused< float > {
...
+              TL_T_REAL                  const *i_val,
               unsigned int                      i_m,
               unsigned int                      i_n,
               unsigned int                      i_k,
               unsigned int                      i_ldA,
               unsigned int                      i_ldB,
               unsigned int                      i_ldC,
-              float                             i_alpha,
-              float                             i_beta,
+              double                            i_alpha,
```
Comment (alheinecke) [inline src/data/MmXsmmFused.hpp]: That should/could be TL_T_REAL

File: src/data/TernaryXsmm.hpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
+/**
+ * @file This file is part of EDGE.
+ *
+ * @author Alexander Breuer (anbreuer AT ucsd.edu)
```
Comment (alheinecke) [inline src/data/TernaryXsmm.hpp]: see comments on binary

File: src/data/UnaryXsmm.hpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
+/**
+ * @file This file is part of EDGE.
+ *
+ * @author Alexander Breuer (anbreuer AT ucsd.edu)
```
Comment (alheinecke) [inline src/data/UnaryXsmm.hpp]: dito

File: src/data/UnaryXsmm.hpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      memset( &unary_param, 0, sizeof(libxsmm_meltw_unary_param) );
+
+      unary_param.in.primary    = (void*)i_in;
+      unary_param.out.primary   = (void*)io_o0;
+      unary_param.out.secondary = (void*)io_o1;
+
+      u_kernels[i_group][i_entry]( &unary_param );
+    }
+
+    void execute( const unsigned short i_group,
```
Comment (alheinecke) [inline src/data/UnaryXsmm.hpp:149]: perhaps a different name to distinguish that we using secondary etc. args?

File: src/data/XsmmUtils.hpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
+/**
+ * @file This file is part of EDGE.
+ *
+ * @author Alexander Breuer (anbreuer AT ucsd.edu)
```
Comment (alheinecke) [inline src/data/XsmmUtils.hpp]: dito

**Review by alheinecke — COMMENTED** [review]
Looks good. Some nits are pointed out.

When reviewing, I saw that AderDG class uses for M=TL_N_MDS_EL*TL_N_CRS; which is fine. I was wondering if we use the same for the kernel if we could collapse more code?

I think we need to do some perf testing on x86 to get some of this decided, so I think for now it's very good that you have PP defines for switching back to the old loops.

File: src/data/BinaryXsmm.hpp (hunk 1 above)
Comment (kvoronin-intel) [reply src/data/BinaryXsmm.hpp]: Not sure what's the right copyright, a copy of the previous ones +  a new one on top or just the new one? (did the latter)

File: src/data/MmXsmmFused.hpp (hunk 2 above)
Comment (kvoronin-intel) [reply src/data/MmXsmmFused.hpp]: Fixed.

File: src/data/TernaryXsmm.hpp (hunk 3 above)
Comment (kvoronin-intel) [reply src/data/TernaryXsmm.hpp]: Adjusted

File: src/data/UnaryXsmm.hpp (hunk 4 above)
Comment (kvoronin-intel) [reply src/data/UnaryXsmm.hpp]: Adjusted

File: src/data/XsmmUtils.hpp (hunk 6 above)
Comment (kvoronin-intel) [reply src/data/XsmmUtils.hpp]: Adjusted

File: src/data/UnaryXsmm.hpp (hunk 5 above)
Comment (kvoronin-intel) [reply src/data/UnaryXsmm.hpp:149]: Is there a reason we don't want overloading here?

**Comment by kvoronin-intel** [comment]
> Looks good. Some nits are pointed out.
> 
> When reviewing, I saw that AderDG class uses for M=TL_N_MDS_EL*TL_N_CRS; which is fine. I was wondering if we use the same for the kernel if we could collapse more code?
> 
> I think we need to do some perf testing on x86 to get some of this decided, so I think for now it's very good that you have PP defines for switching back to the old loops.

The reason for this shape in AderDG is that I ported directly what was done in the older commit here:
https://github.com/alheinecke/edge/commit/eb94c38312062e30b26b9885e08ba969f16e84ef

I'm fine with either choice, we could do more collapsing or less collapsing. Leaving TL_N_CRS as m might be more equivalent to the non-TPPed code (which I guess was tuned for performance to put `#pragma omp simd` in the places they are right now) if we vectorize over m in eltwise TPPs in LIBXSMM

**Review by alheinecke — APPROVED** [review]

File: src/data/UnaryXsmm.hpp (hunk 5 above)
Comment (alheinecke) [reply src/data/UnaryXsmm.hpp:149]: yeah... I think we are fine with overloading, this felt we could do something in addition to make it clearer, but let's stay for now with the current state

File: src/data/BinaryXsmm.hpp (hunk 1 above)
Comment (alheinecke) [reply src/data/BinaryXsmm.hpp]: LGTM

**Comment by alheinecke** [comment]
I will merge for now and then we can do performance testing. After the merge I will also update libxsmm to its latest version which now incl. aarch64 SVE support for the fused operators.

**Pull request merged**